README.md
This is a package to be able to simplify the use of socket.io in flutter.

I just replicated this package by extracting from the pub package made available in the packageslibrary. 

Much of the code was created by the itsclicking.com team licensed under MIT License 
So now I decided do some improvements and share to because they do not have a public repository.

# flutter_socket_io  
  
Flutter Socket IO Plugin, supported Android + iOS (iOS installation guide above)

## How to install on iOS

- 1. Copy the folder `example/ios/Runner/SocketObj` to `${PROJECT_ROOT}/ios/Runner/`

- 2. Replace the `${PROJECT_ROOT}/ios/Runner/AppDelegate.m` with `example/ios/Runner/AppDelegate.m`. 
(Notice: **You should merge the old one in your project to merge with the new from this plugin if you have some change on that file**)

- 3. Open `${PROJECT_ROOT}/ios/Podfile`, paste this line `pod 'Socket.IO-Client-Swift', '~> 13.3.0'` before the end of `target 'Runner' do` block

- 4. Run and Enjoy the plugin :)


## Use the plugin
	

**1. Add the following import to your Dart code:**
```dart
import  'package:flutter_socket_io/flutter_socket_io.dart';
```
	
**2. SocketIOManager**: to manage (create/destroy) list of SocketIO 

*Create SocketIO with SocketIOManager*: 
	
```dart
SocketIO socketIO = SocketIOManager().createSocketIO("http://127.0.0.1:3000", 
		namespace: "/chat", 
		query: "userId=21031", 
		socketStatusCallback: _socketStatus
	);  
```

Destroy SocketIO with socketIOManager:
		
```dart
SocketIOManager().destroySocket(socketIO);  
```
    
**3. SocketIO**:

*Get Id (Url + Namespace) of the socket*
```dart
String getId();
```
</br>
</br>

*Create a new socket and connects the client*
```dart
Future<void> connect();
```
</br>
</br>

*Init socket before doing anything with socket*  
```dart
Future<void> init();
```
</br>
</br>

*Subscribe to a channel with a callback*  
 ```
 Future<void> subscribe(String event, Function callback); 
```dart
</br>
</br>

*Unsubscribe from a channel. When no callback is provided, unsubscribe all subscribers of the channel. Otherwise, unsubscribe only the callback passed in*  
 
```dart
Future<void> unSubscribe(String event, [Function callback]); 
```
</br>
</br>

*Send a message via a channel (i.e. event, *the native code will convert string message to JsonObject before sending*)*  
```dart
Future<void> sendMessage(String event, dynamic message, [Function callback]);
```
</br>
</br>

*Disconnect from the socket*  
```dart
Future<void> disconnect(); 
```
</br>
</br>

*Unsubscribe all subscribers from all channels*  
```dart
Future<void> unSubscribesAll();
```
</br>
</br>

**4. Example:**
[Link](https://pub.dartlang.org/packages/flutter_socket_io#-example-tab-)
  
```dart
SocketIO socketIO;
_connectSocket01() { 
	//update your domain before using  
	 socketIO = SocketIOManager().createSocketIO("http://127.0.0.1:3000", "/chat", query: "userId=21031", socketStatusCallback: _socketStatus); 

	//call init socket before doing anything 
	socketIO.init(); 

	//subscribe event
	socketIO.subscribe("socket_info", _onSocketInfo); 

	//connect socket 
	socketIO.connect(); 
}

_socketStatus(dynamic data) { 
	print("Socket status: " + data); 
}

_subscribes() { 
	if (socketIO != null) { 
		socketIO.subscribe("chat_direct", _onReceiveChatMessage); 
	} 
}

void _onReceiveChatMessage(dynamic message) { 
	print("Message from UFO: " + message); 
}

void _sendChatMessage(String msg) async { 
	if (socketIO != null) { 
		String jsonData = '{"message":{"type":"Text","content": ${(msg != null && msg.isNotEmpty) ? '"${msg}"' : '"Hello SOCKET IO PLUGIN :))"'},"owner":"589f10b9bbcd694aa570988d","avatar":"img/avatar-default.png"},"sender":{"userId":"589f10b9bbcd694aa570988d","first":"Ha","last":"Test 2","location":{"lat":10.792273999999999,"long":106.6430356,"accuracy":38,"regionId":null,"vendor":"gps","verticalAccuracy":null},"name":"Ha Test 2"},"receivers":["587e1147744c6260e2d3a4af"],"conversationId":"589f116612aa254aa4fef79f","name":null,"isAnonymous":null}'; 
		socketIO.sendMessage("chat_direct", jsonData, _onReceiveChatMessage); 
	}
 }

_destroySocket() { 
	if (socketIO != null) { 
		SocketIOManager().destroySocket(socketIO); 
	} 
}




CHANGELOG.md

## 0.4.4
* Added: - Api break changes... optional param ```[namespace]``` default is ```'/'```

## 0.4.3
* Changed: - init with query option
* Fixed bug: - added forceNew to create socket instance

## 0.4.2
* Fixed bug: checked NULL data received before calling back when using sendMessage function* Fixed bug: checked NULL data received from server before calling back when using sendMessage function

## 0.4.1
* Changed: changed android dependency from "com.github.nkzawa:socket.io-client:0.3.0" to official socket.io "io.socket:socket.io-client:1.0.0"

## 0.4.0
* Added: The guide for installing this plugin on iOS 
* Added: MIT LICENSE

## 0.3.8
* Fixed bugs.

## 0.3.7
* Updated documentation

## 0.3.3
* Fixed bugs: iOS don't create new SocketManger (with the same domain) if it already existed.

## 0.3.2
* Fixed bugs.

## 0.3.0
* Supported iOS (Testing).

## 0.2.0
* Added function: ```init()``` socket (please call this function before doing anything with socket)

## 0.1.1
* Refactor code.

## 0.1.0
* Fixed bugs.

## 0.0.21
* Fixed bugs.

## 0.0.20
* Fixed bugs.

## 0.0.19
* Fixed bugs.

## 0.0.18
* Fixed bugs.

## 0.0.17
* Set transports = new String[]{WebSocket.NAME};

## 0.0.16
* Fixed bugs.

## 0.0.15
* Fixed bugs.
* Added 'String query' param ```[optional]``` when calling connect function

## 0.0.14
* Fixed bugs.

## 0.0.12
* Fixed bugs.

## 0.0.11
* Fixed bugs.
* Added features: supported calling connect/subscribes/unSubscribes/sendMessage with Flutter's callback Function.
* Supported OS:
    + Android
    + iOS: COMING SOON

## 0.0.8
* Initial release.
* Supported:
    + Android
    + iOS: coming soon


lib/flutter_socket_io.dart
import 'dart:async';
import 'dart:convert';

import 'package:flutter/services.dart';

class SocketIO {

  String _domain;
  String _namespace;
  String _query;
  Function _statusCallback;
  String _statusCallbackName;
  MethodChannel _channel;

  Map<String, CallbackFunctions> _callbacks;

  ///domain: domain url
  ///namespace: just for iOS
  ///socketStatusCallback [optional]: the status of socket 
  ///[connect/disconnect/reconnect/...] will be sent into this function
  SocketIO(MethodChannel channel, String domain, String namespace, {String query, Function socketStatusCallback}) {
    _channel = channel;
    _domain = domain;
    _namespace = namespace;
    _query = query;

    _callbacks = new Map();
    _statusCallback = socketStatusCallback;
    _statusCallbackName = _parserFunctionName(socketStatusCallback);
  }

  Future<dynamic> handlerMethodCall(String event, String callbackFuncName, dynamic arguments) {
     if (event != null && event.isNotEmpty) {
       if (event == _statusCallbackName && _statusCallback != null) {
         _statusCallback(arguments);
       } else {
         CallbackFunctions functions = _callbacks[event];
         if (functions != null) {
           SocketIOFunction f = functions.getFunctionByName(callbackFuncName);
           if (f != null && f.function != null) {
             print("CALLLING FUNCTION: " + f.functionName);
             f.function(arguments);
             return null;
           }
         }
       }
    }
    return null;
  }

  _clearAll() {
    if(_callbacks != null) {
      _callbacks.clear();
    }
  }

  /// Get Id (Url + Namespace) of the socket
  String getId() {
    if (_domain != null) {
      return _domain + (_namespace != null ? _namespace : "");
    }
    return null;
  }

  /// Init socket before doing anything with socket
  Future<void> init({String query}) async {
    if (query != null) {
      _query = query;
    }

    await _channel.invokeMethod(MethodCallName.SOCKET_INIT, {
      MethodCallArgumentsName.SOCKET_DOMAIN: _domain,
      MethodCallArgumentsName.SOCKET_NAME_SPACE: _namespace,
      MethodCallArgumentsName.SOCKET_QUERY: _query,
      MethodCallArgumentsName.SOCKET_CALLBACK: _statusCallbackName
    });
  }

  /// Create a new socket and connects the client
  Future<void> connect() async {
    await _channel.invokeMethod(MethodCallName.SOCKET_CONNECT, {
      MethodCallArgumentsName.SOCKET_DOMAIN: _domain,
      MethodCallArgumentsName.SOCKET_NAME_SPACE: _namespace,
    });
  }

  /// Subscribe to a channel with a callback
  Future<void> subscribe(String event, Function callback) async {
    if (event != null && event.isNotEmpty) {
      CallbackFunctions functions = _callbacks[event];
      SocketIOFunction f;

      if (functions == null) {
        functions = new CallbackFunctions();
      }

      if (callback != null) {
        f = functions.addFunction(callback);
      }

      _callbacks[event] = functions;

      var subscribes = new Map<String, String>();
      subscribes.putIfAbsent(event, () => f == null ? "" : f.functionName);

      await _channel.invokeMethod(MethodCallName.SOCKET_SUBSCRIBES, {
        MethodCallArgumentsName.SOCKET_DOMAIN: _domain,
        MethodCallArgumentsName.SOCKET_NAME_SPACE: _namespace,
        MethodCallArgumentsName.SOCKET_DATA: json.encode(subscribes),
      });
    }
  }

  /// Unsubscribe from a channel
  ///
  /// When no callback is provided, unsubscribe all subscribers of the channel. Otherwise, unsubscribe only the callback passed in
  Future<void> unSubscribe(String event, [Function callback]) async {

    if (event != null && event.isNotEmpty) {

      CallbackFunctions callbackFunctions = _callbacks[event];

      if (callbackFunctions != null) {

        callbackFunctions.remove(callback);

        if(callbackFunctions.functions.length < 1) {
          _callbacks.remove(event);
        } else {
          _callbacks[event] = callbackFunctions;
        }

      }

      SocketIOFunction f = new SocketIOFunction(callback);
      var unSubscribes = new Map<String, String>();
      unSubscribes.putIfAbsent(event, () => f == null ? "" : f.functionName);
      await _channel.invokeMethod(MethodCallName.SOCKET_UNSUBSCRIBES, {
        MethodCallArgumentsName.SOCKET_DOMAIN: _domain,
        MethodCallArgumentsName.SOCKET_NAME_SPACE: _namespace,
        MethodCallArgumentsName.SOCKET_DATA: json.encode(unSubscribes),
      });
    }
  }

  /// Send a message via a channel (i.e. event)
  Future<void> sendMessage(String event, dynamic message, [Function callback]) async {
    if (event != null && event.isNotEmpty) {
      CallbackFunctions functions = _callbacks[event];
      SocketIOFunction f;

      if (functions == null) {
        functions = new CallbackFunctions();
      }

      if (callback != null) {
        f = functions.addFunction(callback);
      }

      _callbacks[event] = functions;

      await _channel.invokeMethod(MethodCallName.SOCKET_SEND_MESSAGE, {
        MethodCallArgumentsName.SOCKET_DOMAIN: _domain,
        MethodCallArgumentsName.SOCKET_NAME_SPACE: _namespace,
        MethodCallArgumentsName.SOCKET_EVENT: event,
        MethodCallArgumentsName.SOCKET_MESSAGE: message,
        MethodCallArgumentsName.SOCKET_CALLBACK: f == null ? "" : f.functionName
      });
    }
  }

  /// Disconnect from the socket
  Future<void> disconnect() async {
    await _channel.invokeMethod(MethodCallName.SOCKET_DISCONNECT, {
      MethodCallArgumentsName.SOCKET_DOMAIN: _domain,
      MethodCallArgumentsName.SOCKET_NAME_SPACE: _namespace,
    });
  }

  /// Destroy the socket and cleanup all memory usage
  Future<void> destroy() async {
    _clearAll();
    await _channel.invokeMethod(MethodCallName.SOCKET_DESTROY, {
      MethodCallArgumentsName.SOCKET_DOMAIN: _domain,
      MethodCallArgumentsName.SOCKET_NAME_SPACE: _namespace
    });
  }

  /// Unsubscribe all subscribers from all channels
  Future<void> unSubscribesAll() async {
    _clearAll();
    await _channel.invokeMethod(MethodCallName.SOCKET_UNSUBSCRIBES_ALL, {
      MethodCallArgumentsName.SOCKET_DOMAIN: _domain,
      MethodCallArgumentsName.SOCKET_NAME_SPACE: _namespace
    });
  }
}

String _parserFunctionName(Function function) {
  if (function != null) {
    return "FunctionId@${function.hashCode}";
  }
  return null;
}

class SocketIOFunction {
  String _funcName;
  Function _function;

  SocketIOFunction(Function function) {
    _function = function;
    _funcName = _parserFunctionName(function);
  }

  set setFunction(Function function) {
    _function = function;
    _funcName = _parserFunctionName(function);
  }

  String get functionName => _funcName;

  Function get function => _function;
}

class CallbackFunctions {

  List<SocketIOFunction> _functions;

  CallbackFunctions() {
    _functions = new List();
  }

  SocketIOFunction addFunction(Function function) {
    SocketIOFunction f = getFunction(function);
    if (f == null) {
      f = new SocketIOFunction(function);
      _functions.add(f);
    }
    return f;
  }

  SocketIOFunction getFunctionByName(String funcName) {
    if (funcName != null && funcName.isNotEmpty) {
      SocketIOFunction result;
      _functions.forEach((f) {
        if (f != null && f.functionName == funcName) {
          result = f;
          return;
        }
      });
      return result;
    }
    return null;
  }

  SocketIOFunction getFunction(Function function) {
    if (function != null) {
      SocketIOFunction result;
      String funcName = _parserFunctionName(function);
      _functions.forEach((f) {
        if (f != null && funcName != null && f.functionName == funcName) {
          result = f;
          return;
        }
      });
      return result;
    }
    return null;
  }

  bool remove(Function function) {
    if (function != null) {
      String funcName = _parserFunctionName(function);
      SocketIOFunction socketIOFunction;
      _functions.forEach((f) {
        if (f != null && funcName != null && f.functionName == funcName) {
          socketIOFunction = f;
          return;
        }
      });

      return _functions.remove(socketIOFunction);
    }
    return false;
  }

  void removeAllFunctions() {
    if (_functions != null) {
      _functions.clear();
    }
  }

  List<SocketIOFunction> get functions => _functions;
}

class MethodCallArgumentsName {
  static final String SOCKET_DOMAIN = "socketDomain";
  static final String SOCKET_NAME_SPACE = "socketNameSpace";
  static final String SOCKET_CALLBACK = "socketCallback";
  static final String SOCKET_EVENT = "socketEvent";
  static final String SOCKET_MESSAGE = "socketMessage";
  static final String SOCKET_DATA = "socketData";
  static final String SOCKET_QUERY = "socketQuery";
}

class MethodCallName {
  static final String SOCKET_INIT = "socketInit";
  static final String SOCKET_CONNECT = "socketConnect";
  static final String SOCKET_DISCONNECT = "socketDisconnect";
  static final String SOCKET_SUBSCRIBES = "socketSubcribes";
  static final String SOCKET_UNSUBSCRIBES = "socketUnsubcribes";
  static final String SOCKET_UNSUBSCRIBES_ALL = "socketUnsubcribesAll";
  static final String SOCKET_SEND_MESSAGE = "socketSendMessage";
  static final String SOCKET_DESTROY = "socketDestroy";
  static final String SOCKET_DESTROY_ALL = "socketDestroyAll";
}


lib/socket_io_manager.dart
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter_socket_io/flutter_socket_io.dart';

class SocketIOManager {
  static SocketIOManager _instance = new SocketIOManager._internal();
  static const MethodChannel _channel =
      const MethodChannel('flutter_socket_io');
  Map<String, SocketIO> _sockets;

  factory SocketIOManager() {
    return _instance;
  }

  SocketIOManager._internal() {
    _sockets = new Map();
    _channel.setMethodCallHandler(handler);
  }

  Future<dynamic> handler(MethodCall call) {
    //structure of method call: "<socketId>|<event>|<callbackName>"
    if (call.method.contains("|")) {
      dynamic params = _parserMethodCall(call.method);
      print(params);
      if (params != null && params.length > 2) {
        SocketIO socketIO = _getSocketIO(params[0]);
        if (socketIO != null) {
          socketIO.handlerMethodCall(params[1], params[2], call.arguments);
        } else {
          print("NOT FOUND SOCKET ${params[0]}");
        }
      }
    }
    return null;
  }

  String _getSocketId(String domain, String namespace) {
    if (domain != null) {
      return domain + (namespace != null ? namespace : "");
    }
    return null;
  }

  SocketIO _getSocketIO(String socketId) {
    if (socketId != null && socketId.isNotEmpty && _sockets != null) {
      SocketIO result;
      _sockets.forEach((id, socket) {
        if (id == socketId) {
          result = socket;
          return;
        }
      });
      return result;
    }
    return null;
  }

  SocketIO createSocketIO(String domain,
      {String namespace = "/", String query, Function socketStatusCallback}) {
    if (domain == null || domain.isEmpty) {
      print("DOMAIN IS NULL OR EMPTY!");
      return null;
    }
    String socketId = _getSocketId(domain, namespace);
    SocketIO socketIO = _getSocketIO(socketId);

    if (socketIO == null) {
      print("CREATING NEW SOCKET: $socketId");
      socketIO = new SocketIO(_channel, domain, namespace,
          query: query, socketStatusCallback: socketStatusCallback);
      _sockets.putIfAbsent(socketId, () => socketIO);
    }

    return socketIO;
  }

  Future<void> destroySocket(SocketIO socket) async {
    if (socket != null && _sockets != null) {
      _sockets.remove(socket.getId());
      socket.destroy();
      socket = null;
    }
  }

  Future<void> destroyAllSocket() async {
    if (_sockets != null) {
      _sockets.forEach((id, socket) {
        socket.destroy();
      });
      _sockets.clear();
    }
  }

  dynamic _parserMethodCall(String method) {
    if (method != null && method.isNotEmpty && method.contains("|")) {
      return method.split("|");
    }
    return null;
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_socket_io/flutter_socket_io.dart';
import 'package:flutter_socket_io/socket_io_manager.dart';

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or press Run > Flutter Hot Reload in IntelliJ). Notice that the
        // counter didn't reset back to zero; the application is not restarted.
        primarySwatch: Colors.blue,
      ),
      home: new MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  var mTextMessageController = new TextEditingController();
  SocketIO socketIO;
  SocketIO socketIO02;

  @override
  void initState() {
    super.initState();
  }

  _connectSocket01() {
    //update your domain before using
    /*socketIO = new SocketIO("http://127.0.0.1:3000", "/chat",
        query: "userId=21031", socketStatusCallback: _socketStatus);*/
    socketIO = SocketIOManager().createSocketIO("http://127.0.0.1:3000",
        namespace: "/chat",
        query: "userId=21031",
        socketStatusCallback: _socketStatus);

    //call init socket before doing anything
    socketIO.init();

    //subscribe event
    socketIO.subscribe("socket_info", _onSocketInfo);

    //connect socket
    socketIO.connect();
  }

  _connectSocket02() {
    socketIO02 = SocketIOManager().createSocketIO(
      "http://127.0.0.1:3000",
      namespace: "/map",
      query: "userId=21031",
      socketStatusCallback: _socketStatus02,
    );

    //call init socket before doing anything
    socketIO02.init();

    //subscribe event
    socketIO02.subscribe("socket_info", _onSocketInfo02);

    //connect socket
    socketIO02.connect();
  }

  _onSocketInfo(dynamic data) {
    print("Socket info: " + data);
  }

  _socketStatus(dynamic data) {
    print("Socket status: " + data);
  }

  _onSocketInfo02(dynamic data) {
    print("Socket 02 info: " + data);
  }

  _socketStatus02(dynamic data) {
    print("Socket 02 status: " + data);
  }

  _subscribes() {
    if (socketIO != null) {
      socketIO.subscribe("chat_direct", _onReceiveChatMessage);
    }
  }

  _unSubscribes() {
    if (socketIO != null) {
      socketIO.unSubscribe("chat_direct", _onReceiveChatMessage);
    }
  }

  _reconnectSocket() {
    if (socketIO == null) {
      _connectSocket01();
    } else {
      socketIO.connect();
    }
  }

  _disconnectSocket() {
    if (socketIO != null) {
      socketIO.disconnect();
    }
  }

  _destroySocket() {
    if (socketIO != null) {
      SocketIOManager().destroySocket(socketIO);
    }
  }

  void _sendChatMessage(String msg) async {
    if (socketIO != null) {
      String jsonData =
          '{"message":{"type":"Text","content": ${(msg != null && msg.isNotEmpty) ? '"${msg}"' : '"Hello SOCKET IO PLUGIN :))"'},"owner":"589f10b9bbcd694aa570988d","avatar":"img/avatar-default.png"},"sender":{"userId":"589f10b9bbcd694aa570988d","first":"Ha","last":"Test 2","location":{"lat":10.792273999999999,"long":106.6430356,"accuracy":38,"regionId":null,"vendor":"gps","verticalAccuracy":null},"name":"Ha Test 2"},"receivers":["587e1147744c6260e2d3a4af"],"conversationId":"589f116612aa254aa4fef79f","name":null,"isAnonymous":null}';
      socketIO.sendMessage("chat_direct", jsonData, _onReceiveChatMessage);
    }
  }

  void socketInfo(dynamic message) {
    print("Socket Info: " + message);
  }

  void _onReceiveChatMessage(dynamic message) {
    print("Message from UFO: " + message);
  }

  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed(
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  void _showToast() {
    _sendChatMessage(mTextMessageController.text);
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return new Scaffold(
      appBar: new AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: new Text(widget.title),
      ),
      body: new Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: new Column(
          // Column is also layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug paint" (press "p" in the console where you ran
          // "flutter run", or select "Toggle Debug Paint" from the Flutter tool
          // window in IntelliJ) to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            new RaisedButton(
              child: const Text('CONNECT  SOCKET 01',
                  style: TextStyle(color: Colors.white)),
              color: Theme.of(context).accentColor,
              elevation: 0.0,
              splashColor: Colors.blueGrey,
              onPressed: () {
                _connectSocket01();
//                _sendChatMessage(mTextMessageController.text);
              },
            ),
            new RaisedButton(
              child: const Text('CONNECT SOCKET 02',
                  style: TextStyle(color: Colors.white)),
              color: Theme.of(context).accentColor,
              elevation: 0.0,
              splashColor: Colors.blueGrey,
              onPressed: () {
                _connectSocket02();
//                _sendChatMessage(mTextMessageController.text);
              },
            ),
            new RaisedButton(
              child: const Text('SEND MESSAGE',
                  style: TextStyle(color: Colors.white)),
              color: Theme.of(context).accentColor,
              elevation: 0.0,
              splashColor: Colors.blueGrey,
              onPressed: () {
                _showToast();
//                _sendChatMessage(mTextMessageController.text);
              },
            ),
            new RaisedButton(
              child: const Text('SUBSCRIBES',
                  style: TextStyle(color: Colors.white)),
              color: Theme.of(context).accentColor,
              elevation: 0.0,
              splashColor: Colors.blueGrey,
              onPressed: () {
                _subscribes();
//                _sendChatMessage(mTextMessageController.text);
              },
            ),
            new RaisedButton(
              child: const Text('UNSUBSCRIBES',
                  style: TextStyle(color: Colors.white)),
              color: Theme.of(context).accentColor,
              elevation: 0.0,
              splashColor: Colors.blueGrey,
              onPressed: () {
                _unSubscribes();
//                _sendChatMessage(mTextMessageController.text);
              },
            ),
            new RaisedButton(
              child: const Text('RECONNECT',
                  style: TextStyle(color: Colors.white)),
              color: Theme.of(context).accentColor,
              elevation: 0.0,
              splashColor: Colors.blueGrey,
              onPressed: () {
                _reconnectSocket();
//                _sendChatMessage(mTextMessageController.text);
              },
            ),
            new RaisedButton(
              child: const Text('DISCONNECT',
                  style: TextStyle(color: Colors.white)),
              color: Theme.of(context).accentColor,
              elevation: 0.0,
              splashColor: Colors.blueGrey,
              onPressed: () {
                _disconnectSocket();
//                _sendChatMessage(mTextMessageController.text);
              },
            ),
            new RaisedButton(
              child:
                  const Text('DESTROY', style: TextStyle(color: Colors.white)),
              color: Theme.of(context).accentColor,
              elevation: 0.0,
              splashColor: Colors.blueGrey,
              onPressed: () {
                _destroySocket();
//                _sendChatMessage(mTextMessageController.text);
              },
            ),
          ],
        ),
      ),
      floatingActionButton: new FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: new Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}
